### PR TITLE
feat(vue-app): support configurable features

### PR DESCRIPTION
**Pull request by pimlie**
~~This PR is only a draft because we didn't discuss this yet before. Feature-wise I think its complete, maybe just need to add some tests.~~

## Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

The Nuxt Vue-app has quite a bit of features which until this PR were _always_ included, whether you used them or not. This PR defines a list of features which are used to either include or omit the corresponding code in the Vue-app templates. This makes it possible for a user to further optimize their bundle sizes.

Here is the list of features which are toggleable:
```js
  features: {
    router: true, // not implemented
    store: true,
    layouts: true,
    meta: true,
    middleware: true,
    transitions: true,
    deprecations: true,
    validate: true,
    asyncData: true,
    fetch: true,
    clientOnline: true,
    clientPrefetch: true,
    clientUseUrl: false, // this is a bit of an odd one, but using URL should eg be ok for modern mode already
    componentAliases: true,
    componentClientOnly: true
  }
```

I've included a copy of the hello world example app with all (supported) features disabled.

App bundle size of Hello world app
image

App bundle size of Minimal-features Hello world app:
image
image

This PR was also partially given in by the benchmark discussions on discord. You could argue that disabling all these settings is not very usefull and I will probably agree with that. But giving users the possiblity to further optimize their bundle sizes is a nice feature I think and worth the extra work with regards to maintaining the vue-app templates.

Disabling features is considered highly experimental and for advanced users only.

Note: The above bundle size of the Hello world app has already improved slightly compared to the current dev branch. This is mainly due to some optimisations in `./utils.js`. Eg `urlJoin` is/was always included in the app bundle but thats only used in `server.js`

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [ ] All new and existing tests are passing.

### CODE REVIEW

File: packages/builder/src/builder.js [hunk 1]
```diff
@@ -232,6 +232,14 @@ export default class Builder {
       this.resolveMiddleware(templateContext)
     ])
 
+    if (this.options.dev && this.options.build.indicator) {
```
Comment (pi0) [inline packages/builder/src/builder.js]: indicator will be auto disabled for non dev with options.js

File: packages/vue-app/src/index.js [hunk 2]
```diff
@@ -8,15 +8,15 @@ export const template = {
...
     'index.js',
-    'middleware.js',
+    // 'middleware.js',
     'router.js',
     'router.scrollBehavior.js',
     'server.js',
     'utils.js',
     'empty.js',
-    'components/nuxt-build-indicator.vue',
+    // 'components/nuxt-build-indicator.vue',
```
Comment (pi0) [inline packages/vue-app/src/index.js]: For long term: I wanted to export a function instead that takes builder instance to dynamically provide template. Refactoring logic of feature related options here.

File: packages/vue-app/template/App.js [hunk 3]
```diff
@@ -67,10 +83,12 @@ export default {
     // add to window so we can listen when ready
     if (process.client) {
       window.<%= globals.nuxt %> = <%= (globals.nuxt !== '$nuxt' ? 'window.$nuxt = ' : '') %>this
+      <% if (features.client.online) { %>
```
Comment (pi0) [inline packages/vue-app/template/App.js]: Funny fact: As a nuxt developer, wasn't aware we have it 😂😂😂

File: packages/vue-renderer/src/renderers/ssr.js [hunk 4]
```diff
@@ -83,15 +83,18 @@ export default class SSRRenderer extends BaseRenderer {
...
     // Inject head meta
-    const m = renderContext.meta.inject()
-    let HEAD =
-      m.title.text() +
-      m.meta.text() +
-      m.link.text() +
-      m.style.text() +
-      m.script.text() +
-      m.noscript.text()
+    const m = renderContext.meta && renderContext.meta.inject()
```
Comment (pi0) [inline packages/vue-renderer/src/renderers/ssr.js]: Can we also rename m to meta for more readability?

File: packages/vue-app/template/App.js (hunk 3 above)
Comment (manniL) [reply packages/vue-app/template/App.js]: It came together with smart prefetching :relaxed:

**Comment by manniL** [comment]
Amazing idea :+1:

File: packages/vue-app/template/index.js [hunk 5]
```diff
@@ -1,7 +1,7 @@
 import Vue from 'vue'
-import Meta from 'vue-meta'
-import ClientOnly from 'vue-client-only'
-import NoSsr from 'vue-no-ssr'
+<% if (features.meta) { %>import Meta from 'vue-meta'<% } %>
+<% if (features.components.clientOnly) { %>import ClientOnly from 'vue-client-only'<% } %>
+<% if (features.components.noSsr) { %>import NoSsr from 'vue-no-ssr'<% } %>
```
Comment (manniL) [inline packages/vue-app/template/index.js]: Doesn't fit `noSsr` under `deprecations`?

Comment (pimlie) [reply packages/vue-app/template/index.js]: Yes indeed!

**Comment by pimlie** [comment]
Added the suggestions and implemented store/layout features. Also had some wrong/old features in the nuxt.config of the minimal example, so size is even smaller now (image of app bundle size is updated).

**Comment by pi0** [comment]
Really nice enhancement. I wish we could also support disabling router. But seems not an easy task for initial PR.

**Comment by pimlie** [comment]
I looked into making the router configurable as well but almost every function in `client.js` makes use of `from` / `to` params and it seems just to much integrated into all the logic. Dont think it will be an easy task to support disabling the router in the current code. A complete rewrite of client.js with optional router disabling in mind will probably be faster.

File: packages/vue-app/template/client.js [hunk 6]
```diff
@@ -598,42 +653,62 @@ function addHotReload($component, depth) {
...
+    }
     <% } %>
+
     callMiddleware.call(this, Components, context)
     .then(() => {
+      <% if (features.layouts) { %>
       // If layout changed
-      if (depth !== 0) return Promise.resolve()
+      if (depth !== 0) {
+        return true
```
Comment (pimlie) [inline packages/vue-app/template/client.js]: Just a heads-up about this change (and some more below), returning in a thenable chain should always result in a promise.

Try eg: 
```js
new Promise(r => setTimeout(r, 1000)).then(() => true).then((v) => console.log('Done', v))
```

**Review by TheAlexLichter — DISMISSED** [review]

**Comment by Atinux** [comment]
Great work @pimlie, I had this in mind for Nuxt 3.

Ideally, the features enable themselves when used (when detecting the directory associated or defined in the config). Also, I am afraid that some module may not work if the user disable some features 🤔

**Comment by pimlie** [comment]
> Ideally, the features enable themselves when used (when detecting the directory associated or defined in the config).

Well, yes/no. I can imagine there are also times you just want to disable a feature regardless whether you have certain files in your project. Eg with the current store implementation I did you can have a store module in your project, but if you toggle the feature to false it will still be disabled.

> Also, I am afraid that some module may not work if the user disable some features 🤔

True, for the moment thats really up to the user to think about. Thats why I mentioned _experimental and for advanced users only_. 

I dont think we should launch this pr as a great new feature anyway, not as long as we cant provide a router feature. So I would propose to do a silent launch, learn from it and then improve by maybe taking the approach you had in mind for v3

**Comment by manniL** [comment]
> Well, yes/no. I can imagine there are also times you just want to disable a feature regardless whether you have certain files in your project. Eg with the current store implementation I did you can have a store module in your project, but if you toggle the feature to false it will still be disabled.

Definitely, the user prefence should take precedence! But disabling/enabling features if certain conditions are present and the user didn't set the config otherwise would be nice.

File: examples/minimal-features/nuxt.config.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    validate: false,
+    asyncData: false,
+    fetch: false,
+    client: {
+      online: false,
+      prefetch: false,
+      useUrl: true
+    },
+    components: {
+      aliases: false,
```
Comment (pi0) [inline examples/minimal-features/nuxt.config.js]: what do you think about flattened options like `compoentAliases`? This way we don't need to always worry about falsy value in the path (Like `components: false`)

**Review by pi0 — DISMISSED** [review]
A REALLY REALLY nice improvement to support minimal builds for nuxt. Can't wait to also have router disabling support. With nuxt 3 we can have nuxt to be used like:

- Create `app.vue`
- `npx nuxt dev`
- TADA! A fully minimal/universal/efficient vue app is ready!

I think we should be better to merge this ASAP to avoid merge conflicts.

**Comment by pimlie** [comment]
The features config is now a flat object.

Another feature we could add is `errorHandling`. I noticed that after merging the dev branch the app size had grown again due to the changes of the loop prevention for the error component (see latest image in start post). Tried to do that quickly but thats also quite a big change.

@clarkdo Would it be possible somehow to instruct webpack to always create a separate chunk for the complete Nuxt vue-app? I've copied the size limit test and used the `unicode-base` fixture as a size test for a minimal app, but I think for the minimal features it would be much more interesting if we could just test the vue-app bundle. I also lowered the allowed threshold to 1%, as big changes to the vue-app should probably be hidden behind new feature flags.

**Comment by clarkdo** [comment]
@pimlie If you mean splitting the chunk only for testing purpose, I think it's a good idea. You can implement it by adding a `cacheGroup` and filtering files inside `.nuxt`(buildDir).

File: packages/vue-app/template/client.js [hunk 8]
```diff
@@ -599,42 +654,62 @@ function addHotReload($component, depth) {
...
+      }
+
       let layout = Component.options.layout || 'default'
       if (typeof layout === 'function') {
         layout = layout(context)
       }
-      if (this.layoutName === layout) return Promise.resolve()
+      if (this.layoutName === layout) {
+        return true
```
Comment (pi0) [inline packages/vue-app/template/client.js]: Does `true` value means anything special? Before it was resolving to undefined.

Comment (pimlie) [reply packages/vue-app/template/client.js]: No, actually not. Have removed that

**Review by pi0 — DISMISSED** [review]

**Comment by pimlie** [comment]
@clarkdo Thanks, could you please verify if I implemented it correctly?

And I would appreciate your/anyone's help with resolving the error with unicode-base, dont really understand why that now gives an error? The file exists on the fs and is listed in the manifest, does one of you maybe have any idea why jsdom cant find it? Only change for that test is that I added `modern: 'server'` in nuxt.config.

**Comment by clarkdo** [comment]
@pimlie Looks perfect, you can double verify it by having a look at analyze report.

**Review by pi0 — APPROVED** [review]

**Review by clarkdo — APPROVED** [review]

**Review by TheAlexLichter — APPROVED** [review]

**Pull request merged**